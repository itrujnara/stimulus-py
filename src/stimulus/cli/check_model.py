#!/usr/bin/env python3
"""CLI module for checking model configuration and running initial tests."""

import argparse
import json
import logging
import os

import yaml

from stimulus.data.csv import CsvProcessing
from stimulus.learner.raytune_learner import TuneWrapper as StimulusTuneWrapper
from stimulus.utils.json_schema import JsonSchema
from stimulus.utils.launch_utils import get_experiment, import_class_from_file, memory_split_for_ray_init


def get_args() -> argparse.Namespace:
    """Get the arguments when using from the commandline.

    Returns:
        Parsed command line arguments.
    """
    parser = argparse.ArgumentParser(description="Launch check_model.")
    parser.add_argument("-d", "--data", type=str, required=True, metavar="FILE", help="Path to input csv file.")
    parser.add_argument("-m", "--model", type=str, required=True, metavar="FILE", help="Path to model file.")
    parser.add_argument(
        "-e",
        "--experiment",
        type=str,
        required=True,
        metavar="FILE",
        help="Experiment config file. From this the experiment class name is extracted.",
    )
    parser.add_argument(
        "-c",
        "--config",
        type=str,
        required=True,
        metavar="FILE",
        help="Path to yaml config training file.",
    )
    parser.add_argument(
        "-w",
        "--initial_weights",
        type=str,
        required=False,
        nargs="?",
        const=None,
        default=None,
        metavar="FILE",
        help="The path to the initial weights. These can be used by the model instead of the random initialization.",
    )
    parser.add_argument(
        "--gpus",
        type=int,
        required=False,
        nargs="?",
        const=None,
        default=None,
        metavar="NUM_OF_MAX_GPU",
        help="Use to limit the number of GPUs ray can use. This might be useful on many occasions, especially in a cluster system.",
    )
    parser.add_argument(
        "--cpus",
        type=int,
        required=False,
        nargs="?",
        const=None,
        default=None,
        metavar="NUM_OF_MAX_CPU",
        help="Use to limit the number of CPUs ray can use. This might be useful on many occasions, especially in a cluster system.",
    )
    parser.add_argument(
        "--memory",
        type=str,
        required=False,
        nargs="?",
        const=None,
        default=None,
        metavar="MAX_MEMORY",
        help="Ray can have a limiter on the total memory it can use. This might be useful on many occasions, especially in a cluster system.",
    )
    parser.add_argument(
        "-n",
        "--num_samples",
        type=int,
        required=False,
        nargs="?",
        const=3,
        default=3,
        metavar="NUM_SAMPLES",
        help="Number of samples for tuning. Overwrites tune.tune_params.num_samples in config.",
    )
    parser.add_argument(
        "--ray_results_dirpath",
        type=str,
        required=False,
        nargs="?",
        const=None,
        default=None,
        metavar="DIR_PATH",
        help="Location where ray_results output dir should be written. If None, uses ~/ray_results.",
    )
    parser.add_argument(
        "--debug_mode",
        action="store_true",
        help="Activate debug mode for tuning. Default false, no debug.",
    )

    return parser.parse_args()


def main(
    data_path: str,
    model_path: str,
    experiment_config: str,
    config_path: str,
    initial_weights_path: str | None = None,
    gpus: int | None = None,
    cpus: int | None = None,
    memory: str | None = None,
    num_samples: int = 3,
    ray_results_dirpath: str | None = None,
    *,
    debug_mode: bool = False,
) -> None:
    """Run the main model checking pipeline.

    Args:
        data_path: Path to input data file.
        model_path: Path to model file.
        experiment_config: Path to experiment config.
        config_path: Path to training config.
        initial_weights_path: Optional path to initial weights.
        gpus: Maximum number of GPUs to use.
        cpus: Maximum number of CPUs to use.
        memory: Maximum memory to use.
        num_samples: Number of samples for tuning.
        ray_results_dirpath: Directory for ray results.
        debug_mode: Whether to run in debug mode.
    """
    # Load experiment config
    with open(experiment_config) as in_json:
        exp_config = json.load(in_json)

    # Initialize json schema and experiment class
    schema = JsonSchema(exp_config)
    initialized_experiment_class = get_experiment(schema.experiment)
    model_class = import_class_from_file(model_path)

    # Update tune config
    updated_tune_conf = "check_model_modified_tune_config.yaml"
    with open(config_path) as conf_file, open(updated_tune_conf, "w") as new_conf:
        user_tune_config = yaml.safe_load(conf_file)
        user_tune_config["tune"]["tune_params"]["num_samples"] = num_samples

        if user_tune_config["tune"]["scheduler"]["name"] == "ASHAScheduler":
            user_tune_config["tune"]["scheduler"]["params"]["max_t"] = 1
            user_tune_config["tune"]["scheduler"]["params"]["grace_period"] = 1
            user_tune_config["tune"]["step_size"] = 1
        elif user_tune_config["tune"]["scheduler"]["name"] == "FIFOScheduler":
            user_tune_config["tune"]["run_params"]["stop"]["training_iteration"] = 1

        if initial_weights_path is not None:
            user_tune_config["model_params"]["initial_weights"] = os.path.abspath(initial_weights_path)

        yaml.dump(user_tune_config, new_conf)

    # Process CSV data
    csv_obj = CsvProcessing(initialized_experiment_class, data_path)
    downsampled_csv = "downsampled.csv"

    if "split" not in csv_obj.check_and_get_categories():
        config_default = {"name": "RandomSplitter", "params": {"split": [0.5, 0.5, 0.0]}}
        csv_obj.add_split(config_default)

    csv_obj.save(downsampled_csv)

    # Initialize ray
    object_store_mem, mem = memory_split_for_ray_init(memory)
    ray_results_dirpath = None if ray_results_dirpath is None else os.path.abspath(ray_results_dirpath)

    # Create and run learner
    learner = StimulusTuneWrapper(
        updated_tune_conf,
        model_class,
        downsampled_csv,
        initialized_experiment_class,
        max_gpus=gpus,
        max_cpus=cpus,
        max_object_store_mem=object_store_mem,
        max_mem=mem,
        ray_results_dir=ray_results_dirpath,
        _debug=debug_mode,
    )

    grid_results = learner.tune()

    # Check results
    logger = logging.getLogger(__name__)
    for i, result in enumerate(grid_results):
        if not result.error:
            logger.info("Trial %d finished successfully with metrics %s.", i, result.metrics)
        else:
            raise TypeError(f"Trial {i} failed with error {result.error}.")


def run() -> None:
    """Run the model checking script."""
    args = get_args()
    main(
        args.data,
        args.model,
        args.experiment,
        args.config,
        args.initial_weights,
        args.gpus,
        args.cpus,
        args.memory,
        args.num_samples,
        args.ray_results_dirpath,
        debug_mode=args.debug_mode,
    )


if __name__ == "__main__":
    run()
