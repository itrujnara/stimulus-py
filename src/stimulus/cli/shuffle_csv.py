#!/usr/bin/env python3
"""CLI module for shuffling CSV data files."""

import argparse
import json
import os

from stimulus.data.data_handlers import CsvProcessing
from stimulus.utils.launch_utils import get_experiment


def get_args() -> argparse.Namespace:
    """Get the arguments when using from the commandline.

    Returns:
        Parsed command line arguments.
    """
    parser = argparse.ArgumentParser(description="Shuffle rows in a CSV data file.")
    parser.add_argument(
        "-c",
        "--csv",
        type=str,
        required=True,
        metavar="FILE",
        help="The file path for the csv containing all data",
    )
    parser.add_argument(
        "-j",
        "--json",
        type=str,
        required=True,
        metavar="FILE",
        help="The json config file that hold all parameter info",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        required=True,
        metavar="FILE",
        help="The output file path to write the noised csv",
    )

    return parser.parse_args()


def main(data_csv: str, config_json: str, out_path: str) -> None:
    """Shuffle the data and split it according to the default split method.

    Args:
        data_csv: Path to input CSV file.
        config_json: Path to config JSON file.
        out_path: Path to output shuffled CSV.

    TODO major changes when this is going to select a given shuffle method and integration with split.
    """
    # open and read Json, just to extract the experiment name, so all other fields are scratched
    config = None
    with open(config_json) as in_json:
        tmp = json.load(in_json)
        config = tmp
        # add fake transform informations
        config["transform"] = "shuffle (special case)"

    # write the config modified, this will be associated to the shuffled data. TODO better solution to renaming like this
    modified_json = os.path.splitext(os.path.basename(data_csv))[0].split("-split")[0] + "-shuffled-experiment.json"
    with open(modified_json, "w") as out_json:
        json.dump(config, out_json)

    # initialize the experiment class
    exp_obj = get_experiment(config["experiment"])

    # initialize the csv processing class, it open and reads the csv in automatic
    csv_obj = CsvProcessing(exp_obj, data_csv)

    # shuffle the data with a default seed. TODO get the seed for the config if and when that is going to be set there.
    csv_obj.shuffle_labels(seed=42)

    # save the modified csv
    csv_obj.save(out_path)


def run() -> None:
    """Run the CSV shuffling script."""
    args = get_args()
    main(args.csv, args.json, args.output)


if __name__ == "__main__":
    run()
