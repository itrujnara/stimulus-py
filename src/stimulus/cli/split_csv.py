#!/usr/bin/env python3
"""CLI module for splitting CSV data files."""

import argparse
import json
import logging

from stimulus.data.csv import CsvProcessing
from stimulus.utils.launch_utils import get_experiment


def get_args() -> argparse.Namespace:
    """Get the arguments when using from the commandline."""
    parser = argparse.ArgumentParser(description="Split a CSV data file.")
    parser.add_argument(
        "-c",
        "--csv",
        type=str,
        required=True,
        metavar="FILE",
        help="The file path for the csv containing all data",
    )
    parser.add_argument(
        "-j",
        "--json",
        type=str,
        required=True,
        metavar="FILE",
        help="The json config file that hold all parameter info",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        required=True,
        metavar="FILE",
        help="The output file path to write the noised csv",
    )

    return parser.parse_args()


def main(data_csv: str, config_json: str, out_path: str) -> None:
    """Connect CSV and JSON configuration and handle sanity checks.

    Args:
        data_csv: Path to input CSV file.
        config_json: Path to config JSON file.
        out_path: Path to output split CSV.

    TODO what happens when the user write his own experiment class? how should he do it ? how does it integrates here?
    """
    # open and read Json
    config = {}
    with open(config_json) as in_json:
        config = json.load(in_json)

    # initialize the experiment class
    exp_obj = get_experiment(config["experiment"])

    # initialize the csv processing class, it open and reads the csv in automatic
    csv_obj = CsvProcessing(exp_obj, data_csv)

    # CASE 1: SPLIT in csv, not in json --> keep the split from the csv
    if "split" in csv_obj.check_and_get_categories() and config["split"] is None:
        pass

    # CASE 2: SPLIT in csv and in json --> use the split from the json
    # TODO change this behaviour to do both, maybe
    elif "split" in csv_obj.check_and_get_categories() and config["split"]:
        logging.info("SPLIT present in both csv and json --> use the split from the json")
        csv_obj.add_split(config["split"], force=True)

    # CASE 3: SPLIT nor in csv and or json --> use the default RandomSplitter
    elif "split" not in csv_obj.check_and_get_categories() and config["split"] is None:
        # In case no split is provided, we use the default RandomSplitter
        logging.warning("SPLIT nor in csv and or json --> use the default RandomSplitter")
        # if the user config is None then set to default splitter -> RandomSplitter.
        config_default = {"name": "RandomSplitter", "params": {}}
        csv_obj.add_split(config_default)

    # CASE 4: SPLIT in json, not in csv --> use the split from the json
    else:
        csv_obj.add_split(config["split"], force=True)

    # save the modified csv
    csv_obj.save(out_path)


def run() -> None:
    """Run the CSV splitting script."""
    args = get_args()
    main(args.csv, args.json, args.output)


if __name__ == "__main__":
    run()
