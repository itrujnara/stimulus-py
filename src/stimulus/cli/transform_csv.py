#!/usr/bin/env python3
"""CLI module for transforming CSV data files."""

import argparse
import json

from stimulus.data.data_handlers import CsvProcessing
from stimulus.utils.launch_utils import get_experiment


def get_args() -> argparse.Namespace:
    """Get the arguments when using from the commandline."""
    parser = argparse.ArgumentParser(description="")
    parser.add_argument(
        "-c",
        "--csv",
        type=str,
        required=True,
        metavar="FILE",
        help="The file path for the csv containing all data",
    )
    parser.add_argument(
        "-j",
        "--json",
        type=str,
        required=True,
        metavar="FILE",
        help="The json config file that hold all parameter info",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        required=True,
        metavar="FILE",
        help="The output file path to write the noised csv",
    )

    return parser.parse_args()


def main(data_csv: str, config_json: str, out_path: str) -> None:
    """Connect CSV and JSON configuration and handle sanity checks.

    This launcher will be the connection between the csv and one json configuration.
    It should also handle some sanity checks.
    """
    # open and read Json
    config = {}
    with open(config_json) as in_json:
        config = json.load(in_json)

    # initialize the experiment class
    exp_obj = get_experiment(config["experiment"])

    # initialize the csv processing class, it open and reads the csv in automatic
    csv_obj = CsvProcessing(exp_obj, data_csv)

    # Transform the data according to what defined in the experiment class and the specifics of the user in the Json
    # in case of no transformation specification so when the config has "augmentation" : None  just save a copy of the original csv file
    if config.get("transform") is not None:
        csv_obj.transform(config["transform"])

    # save the modified csv
    csv_obj.save(out_path)


def run() -> None:
    """Run the CSV transformation script."""
    args = get_args()
    main(args.csv, args.json, args.output)


if __name__ == "__main__":
    run()
