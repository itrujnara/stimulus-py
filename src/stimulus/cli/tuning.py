#!/usr/bin/env python3
"""CLI module for running raytune tuning experiment."""

import argparse
import logging

import yaml
from torch.utils.data import DataLoader

from stimulus.data import handlertorch, loaders
from stimulus.learner import raytune_learner, raytune_parser
from stimulus.utils import launch_utils, yaml_data, yaml_model_schema

logger = logging.getLogger(__name__)


def get_args() -> argparse.Namespace:
    """Get the arguments when using from the commandline.

    Returns:
        Parsed command line arguments.
    """
    parser = argparse.ArgumentParser(description="Launch check_model.")
    parser.add_argument("-d", "--data", type=str, required=True, metavar="FILE", help="Path to input csv file.")
    parser.add_argument("-m", "--model", type=str, required=True, metavar="FILE", help="Path to model file.")
    parser.add_argument(
        "-e",
        "--data_config",
        type=str,
        required=True,
        metavar="FILE",
        help="Path to data config file.",
    )
    parser.add_argument(
        "-c",
        "--model_config",
        type=str,
        required=True,
        metavar="FILE",
        help="Path to yaml config training file.",
    )
    parser.add_argument(
        "-w",
        "--initial_weights",
        type=str,
        required=False,
        nargs="?",
        const=None,
        default=None,
        metavar="FILE",
        help="The path to the initial weights (optional).",
    )


    parser.add_argument(
        "--ray_results_dirpath",
        type=str,
        required=False,
        nargs="?",
        const=None,
        default=None,
        metavar="DIR_PATH",
        help="Location where ray_results output dir should be written. If None, uses ~/ray_results.",
    )

    parser.add_argument(
        "-o",
        "--output",
        type=str,
        required=False,
        nargs="?",
        const="best_model.pt",
        default="best_model.pt",
        metavar="FILE",
        help="The output file path to write the trained model to",
    )

    parser.add_argument(
        "-bm",
        "--best_metrics",
        type=str,
        required=False,
        nargs="?",
        const="best_metrics.csv",
        default="best_metrics.csv",
        metavar="FILE",
        help="The path to write the best metrics to",
    )

    parser.add_argument(
        "-bc",
        "--best_config",
        type=str,
        required=False,
        nargs="?",
        const="best_config.yaml",
        default="best_config.yaml",
        metavar="FILE",
        help="The path to write the best config to",
    )

    parser.add_argument(
        "-bo",
        "--best_optimizer",
        type=str,
        required=False,
        nargs="?",
        const="best_optimizer.pt",
        default="best_optimizer.pt",
        metavar="FILE",
        help="The path to write the best optimizer to",
    )

    parser.add_argument(
        "--tune_run_name",
        type=str,
        required=False,
        nargs="?",
        const=None,
        default=None,
        metavar="CUSTOM_RUN_NAME",
        help="tells ray tune what that the 'experiment_name' aka the given tune_run name should be. This is controlled be the variable name in the RunConfig class of tune. This has two behaviuors: 1 if set the subdir of ray_results is going to be named with this value, 2 the subdir of the above mentioned will also have this value as prefix for the single train dir name. Default None, meaning ray will generate such a name on its own.",
    )

    parser.add_argument(
        "--debug_mode",
        action="store_true",
        help="Activate debug mode for tuning. Default false, no debug.",
    )

    return parser.parse_args()


def main(
    model_path: str,
    data_path: str,
    data_config_path: str,
    model_config_path: str,
    initial_weights: str | None = None,  # noqa: ARG001
    ray_results_dirpath: str | None = None,
    output_path: str | None = None,
    best_optimizer_path: str | None = None,
    best_metrics_path: str | None = None,
    best_config_path: str | None = None,
    *,
    debug_mode: bool = False,
) -> None:
    """Run the main model checking pipeline.

    Args:
        data_path: Path to input data file.
        model_path: Path to model file.
        data_config_path: Path to data config file.
        model_config_path: Path to model config file.
        initial_weights: Optional path to initial weights.
        ray_results_dirpath: Directory for ray results.
        debug_mode: Whether to run in debug mode.
        output_path: Path to write the best model to.
        best_optimizer_path: Path to write the best optimizer to.
        best_metrics_path: Path to write the best metrics to.
        best_config_path: Path to write the best config to.
    """
    with open(data_config_path) as file:
        data_config = yaml.safe_load(file)
        data_config = yaml_data.YamlSubConfigDict(**data_config)

    with open(model_config_path) as file:
        model_config = yaml.safe_load(file)
        model_config = yaml_model_schema.Model(**model_config)

    encoder_loader = loaders.EncoderLoader()
    encoder_loader.initialize_column_encoders_from_config(column_config=data_config.columns)

    model_class = launch_utils.import_class_from_file(model_path)

    ray_config_loader = yaml_model_schema.YamlRayConfigLoader(model=model_config)
    ray_config_model = ray_config_loader.get_config()

    tuner = raytune_learner.TuneWrapper(
        model_config=ray_config_model,
        data_config_path=data_config_path,
        model_class=model_class,
        data_path=data_path,
        encoder_loader=encoder_loader,
        seed=42,
        ray_results_dir=ray_results_dirpath,
        debug=debug_mode,
    )

    grid_results = tuner.tune()

    parser = raytune_parser.TuneParser(grid_results=grid_results)
    parser.save_best_model(output_path=output_path)
    parser.save_best_optimizer(output_path=best_optimizer_path)
    parser.save_best_metrics_dataframe(output_path=best_metrics_path)
    parser.save_best_config(output_path=best_config_path)

def run() -> None:
    """Run the model checking script."""
    args = get_args()
    main(
        data_path=args.data,
        model_path=args.model,
        data_config_path=args.data_config,
        model_config_path=args.model_config,
        initial_weights=args.initial_weights,
        num_samples=args.num_samples,
        ray_results_dirpath=args.ray_results_dirpath,
        debug_mode=args.debug_mode,
    )


if __name__ == "__main__":
    run()
