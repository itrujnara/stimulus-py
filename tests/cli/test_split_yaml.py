import hashlib
import os
import tempfile

import pytest

from src.stimulus.cli.split_yaml import main


# Fixtures
@pytest.fixture
def correct_yaml_path() -> str:
    """Fixture that returns the path to a correct YAML file."""
    return "tests/test_data/titanic/titanic.yaml"


@pytest.fixture
def wrong_yaml_path() -> str:
    """Fixture that returns the path to a wrong YAML file."""
    return "tests/test_data/yaml_files/wrong_field_type.yaml"


# Test cases
test_cases = [
    ("correct_yaml_path", None),
    ("wrong_yaml_path", ValueError),
]


# Tests
@pytest.mark.parametrize("yaml_type, error", test_cases)
def test_split_yaml(request: pytest.FixtureRequest, snapshot, yaml_type: str, error: Exception | None) -> None:
    """Tests the CLI command with correct and wrong YAML files."""
    yaml_path = request.getfixturevalue(yaml_type)
    tmpdir = tempfile.gettempdir()
    if error:
        with pytest.raises(error):
            main(yaml_path, tmpdir)
    else:
        assert main(yaml_path, tmpdir) is None  # this is to assert that the function does not raise any exceptions
        files = os.listdir(tmpdir)
        test_out = [f for f in files if f.startswith("test_")]
        hashes = []
        for f in test_out:
            with open(os.path.join(tmpdir, f)) as file:
                hashes.append(hashlib.md5(file.read().encode()).hexdigest())  # noqa: S324
        assert sorted(hashes) == snapshot # sorted ensures that the order of the hashes does not matter
