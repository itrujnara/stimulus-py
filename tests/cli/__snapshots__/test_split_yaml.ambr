# serializer version: 1
# name: test_split_yaml[correct_yaml_path-None]
  list([
    '455bac9343934e1ff40130ee94d5aa29',
    '5a8a9dd96d15932d28254bde3949d7ea',
    'a66d7aa1817e90ecdc81f02591f50289',
  ])
# ---
